Bug fixes and improvements

samdogg7 opened the pull request:
Hi @wojtek717!

I was searching for a lightweight solution like your's yesterday. This is great, thank you for creating this package!

I started to experiment with your package in my app and caught a couple of issues.

## Permanently hidden `BottomSheetView`

I noticed that the `BottomSheetView` would sometimes get into a state where it was permanently hidden.

Before this PR:

https://github.com/user-attachments/assets/03151a6d-b8a1-4fa5-b177-91aa85cc96fb 

After this PR:

https://github.com/user-attachments/assets/ec543d40-aa8b-44ff-b827-1d7b57f2e3d9

## Selected `Detent`

Added the concept of a `selectedDetent`. This persists the bottom sheet's detent state when switching tabs. Also added support for selecting the initial detent of the bottom sheet, which defaults to the smallest detent.

Before this PR:

https://github.com/user-attachments/assets/cf71ee43-968c-406b-b0fc-263dd2d46564

After this PR:

https://github.com/user-attachments/assets/041fafe5-2a4e-4848-9f4e-655af4244ad6

## "Jumpy" drag gestures

When a user is doing a long drag gesture, sometimes the `translation`'s `height` evaluates to be some very large / small value. In this PR I smooth and clamp the drag gesture, to resolve this.

Before this PR: 

https://github.com/user-attachments/assets/a5f0ca89-da78-4fce-a92d-80d9256c256c

After this PR: 

https://github.com/user-attachments/assets/34a06838-efee-4994-b536-8de8560ea8d4

## Miscellaneous changes

I made `BottomSheetViewConfiguration` conform to `@Observable` and made it `@State` in `BottomSheetView` / `BottomSheet`. You previously had it as a `var` that would be updated by view modifiers. The only reason those changes would work is because the `sheetHeight` was being called frequently enabling the SwiftUI engine to check `BottomSheetViewConfiguration`'s new state.

I noticed you were adding `ViewBuilder` closures as instance properties, such as `@ViewBuilder private let content: () -> Content`, and that you were using `@escaping` in the initializer. This is considered bad practice, there can be unintended side effects. To learn more about this, I recommend reading [this](https://rensbr.eu/blog/swiftui-escaping-closures/) article! You can see the changes I made in `BottomSheet` that fix this.

Lastly, I did some general quality of life fixes like exposing `cornerRadius` for the sheet, naming improvements, and removed some forced view modifiers, such as the forced bottom safe area being ignored.

I may open a second PR with some other optional suggestions, but I wanted to keep my changes limited to what I saw as essential for the package.

samdogg7 commented on Sources/BottomSheet/BottomSheet.swift:
```diff
@@ -7,55 +7,84 @@
             
             if isPresented {
-                BottomSheetView(configuration: configuration, content: sheetContent)
-                    .ignoresSafeAreaEdges(configuration.ignoredEdges)
```
Hmmm I am realizing now that this isn't publicly exposed, I'll re-add

samdogg7 replied:
In [434b17c](https://github.com/wojtek717/bottom-sheet/pull/8/commits/434b17c29607879805070c73ed1666b4363c9a56) I re-added support for ignoring safe area for the bottom sheet. 

In my opinion, `.bottom` should _not_ be the default value. This results `List` content being hidden behind the `TabView` which is typically not behavior people would like. Here are demo videos.

With `.bottom`:

https://github.com/user-attachments/assets/3cbf88dc-ee2c-4857-a740-1c96221e33d8

Without `.bottom`:

https://github.com/user-attachments/assets/2551a1e6-a357-4bdf-85cf-ea8d9c3ff7e4

samdogg7 commented on Sources/BottomSheet/BottomSheet.swift:
```diff
@@ -7,55 +7,84 @@
+        TabView {
+            Tab("Map", systemImage: "map") {
+                Text("Map")
+                    .bottomSheet {
+                        rainbowList
+                    }
+                    .dragIndicatorPresentation(isVisible: true)
+                    .detentsPresentation(detents: [.small, .medium, .large])
```
When `bottomSheet` is initially called, it hasn't received the `detentsPresentation(detents: [.small, .medium, .large])` modifier's changes. As a result, `detents` is initially `[.large]`, causing `selectedDetent` to always be `.large`, even if the developer doesn't desire `.large`. Working on resolving this issue now.

samdogg7 replied:
Fixed in [8854da9](https://github.com/wojtek717/bottom-sheet/pull/8/commits/8854da952d5caf2960ceb7a4e24625c3a150071c). Added support for selecting the initial detent of the bottom sheet, and defaults to the smallest state.

https://github.com/user-attachments/assets/7293527a-a777-4e43-ac3d-1975eb14ecd1

samdogg7 commented:
Moved this PR to draft state - there are a couple of regressions introduced

samdogg7 commented:
I am noticing some bad drag behavior as a result of my `sheetHeight` changes. My assumption is the `onScrollGeometryChange`'s `action` is the source of this issue. 

I added `max` to ensure the bottom sheet wouldn't be permanently hidden:

```swift
if newValue < 0.0 {
    sheetHeight = max(
        minHeight(for: screenHeight),
        sheetHeight + newValue
    )
}
```

But I think the drag gesture is swapping between `minHeight` and `sheetHeight + newValue` causing this bad behavior:

https://github.com/user-attachments/assets/a5f0ca89-da78-4fce-a92d-80d9256c256c

samdogg7 commented:
Fixed the "jumpy" behavior noted in the last [comment](https://github.com/wojtek717/bottom-sheet/pull/8#issuecomment-2440042981) in the latest commit, [d5698a9](https://github.com/wojtek717/bottom-sheet/pull/8/commits/d5698a9fd0bcd1e22054284bc7c24fa87031de4f). This clamps and smooths the drag gestures.

https://github.com/user-attachments/assets/34a06838-efee-4994-b536-8de8560ea8d4

wojtek717 commented:
@samdogg7 Thank you very much for this PR and for your accurate comments. I had some of your corrections on target myself but you preceded me 💪! I will check and test your changes as soon as possible 👀. 
I am glad that my work has found an audience and a willingness to further improve this package.

wojtek717 commented:
@samdogg7 Would you mind creating a cumulative issue briefly listing the fixes you have made? I think it would be useful for us to keep changlelog in an organized way.

samdogg7 commented:
No problem! This is awesome. I can definitely write up a consolidated list of changes and comment them here.

I have a few other suggestions that I've left out of this PR- I think I'll open it separately with you. They are preferential, and not necessary fixes.

samdogg7 commented:
I definitely opened the PR early 😅 Caught some things after posting it

wojtek717 commented:
Thanks, let's focus on this PR and later we can create a list with next improvements/bugs to do.

samdogg7 commented:
@wojtek717 here is the change summary:

### Changelog

**Fixes:**
- Resolved issue with `BottomSheetView` becoming permanently hidden.
- Smoothed and clamped drag gestures to prevent "jumpy" behavior during long drags.

**Enhancements:**
- Introduced `selectedDetent` to persist detent state across tabs and allow setting an initial detent (default: smallest).
- Refactored `BottomSheetViewConfiguration` to conform to `@Observable` and use `@State`, enhancing state consistency.
- Exposed `cornerRadius` customization for the bottom sheet.
- Improved code quality with naming refinements and removal of forced view modifiers (e.g., forced bottom safe area).
- Updated the example of `BottomSheet`.

**Best Practices:**
- Refined handling of `ViewBuilder` closures in `BottomSheet` to avoid unintended side effects by removing `@escaping` in the initializer.

wojtek717 commented:
close #11

wojtek717 commented:
@samdogg7 One last thing, can you pull latest main branch to this pr so GH Action will run?

samdogg7 commented:
@wojtek717 merged `main` into the branch!
